pkg/csi/controller_server.go

```diff
@@ -457,31 +657,111 @@ func getVolumeCapabilityAccessModes(vc []csi.VolumeCapability_AccessMode_Mode) [
 	return vca
 }
 
-func (cs *ControllerServer) waitForPVCState(name string, stateDescription string,
-	predicate func(pvc *corev1.PersistentVolumeClaim) bool) bool {
-	timer := time.NewTimer(timeoutAttachDetach)
-	defer timer.Stop()
-	timeout := timer.C
-
+func (cs *ControllerServer) waitForLHVolumeName(pvcName string) bool {
+	timeoutTimer := time.NewTimer(timeoutAttachDetach)
+	defer timeoutTimer.Stop()
 	ticker := time.NewTicker(tickAttachDetach)
 	defer ticker.Stop()
+
+	timeout := timeoutTimer.C
 	tick := ticker.C
 
 	for {
 		select {
 		case <-timeout:
-			logrus.Warnf("waitForPVCState: timeout while waiting for PVC %s state %s", name, stateDescription)
+			logrus.Warnf("waitForLHVolumeName: timeout while waiting for volume %s to be created", pvcName)
 			return false
 		case <-tick:
-			logrus.Debugf("Polling PVC %s state for %s at %s", name, stateDescription, time.Now().String())
-			existVol, err := cs.coreClient.PersistentVolumeClaim().Get(cs.namespace, name, metav1.GetOptions{})
+			resPVC, err := cs.coreClient.PersistentVolumeClaim().Get(cs.namespace, pvcName, metav1.GetOptions{})
 			if err != nil {
-				logrus.Warnf("waitForPVCState: error while waiting for PVC %s state %s error %s", name, stateDescription, err)
+				logrus.Warnf("waitForLHVolumeName: error while waiting for volume %s to be created: %v", pvcName, err)
+				return false
+			}
+			if resPVC.Spec.VolumeName == "" {
+				logrus.Infof("volumeName is not set for PVC %s, continue to wait", pvcName)
 				continue
 			}
-			if predicate(existVol) {
-				return true
+			return true
+		}
+	}
+}
+
+func (cs *ControllerServer) waitForLHVolumeDetachedorAttachedCorrectly(volName, nodeID string) (detached, attached bool) {
+	timeoutTimer := time.NewTimer(timeoutAttachDetach)
```

Comment:
nit: might worth checking: https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#PollUntilContextTimeout, while there is nothing wrong implementing the logic by ourself too.

Reply:
How about handling the attach/detach separately (with an argument or different functions) and returning the error from the function?

Reply:
Thanks for the suggestion, I rewrite this with `PollUnitContextTimeout`.
Please help to check. Thanks!